### ISSUE TITLE
Switching workspace on focused monitor sometimes also changes workspace on non-focused monitor

### ISSUE THREAD

**Issue by macearl** [hyprwm/Hyprland#2154]
# Description
When switching to another workspace with `hyprctl` and the dispatches `moveworkspacetomonitor` and `workspace` on my focused monitor sometimes the workspace on the non-focused monitor also changes.

I'm using a wrapper script to change my workspaces as i prefer to have the new workspace open on my currently active monitor (the Script is "heavily inspired" by https://github.com/taylor85345/hyprland-dotfiles/blob/master/hypr/scripts/workspace which i found through:    https://github.com/hyprwm/Hyprland/issues/747 )

I'm quite new to hyprland and wayland in general, so if i should test anything or provide more logs i would be more than happy to.

# Steps to reproduce
Create the wrapper script and configure the keybinding:
## config: included logging the output for debug purposes
```
[...]
# Switch workspaces with mainMod + [0-9]
bind = $mainMod, 1, exec, bash -c "/home/myusername/scripts/hyprland_show_workspace_clean.sh 1 >> /home/myusername/monitor.log 2>&1"
bind = $mainMod, 2, exec, bash -c "/home/myusername/scripts/hyprland_show_workspace_clean.sh 2 >> /home/myusername/monitor.log 2>&1"
bind = $mainMod, 3, exec, bash -c "/home/myusername/scripts/hyprland_show_workspace_clean.sh 3 >> /home/myusername/monitor.log 2>&1"
bind = $mainMod, 4, exec, bash -c "/home/myusername/scripts/hyprland_show_workspace_clean.sh 4 >> /home/myusername/monitor.log 2>&1"
bind = $mainMod, 5, exec, bash -c "/home/myusername/scripts/hyprland_show_workspace_clean.sh 5 >> /home/myusername/monitor.log 2>&1"
bind = $mainMod, 6, exec, bash -c "/home/myusername/scripts/hyprland_show_workspace_clean.sh 6 >> /home/myusername/monitor.log 2>&1"
bind = $mainMod, 7, exec, bash -c "/home/myusername/scripts/hyprland_show_workspace_clean.sh 7 >> /home/myusername/monitor.log 2>&1"
bind = $mainMod, 8, exec, bash -c "/home/myusername/scripts/hyprland_show_workspace_clean.sh 8 >> /home/myusername/monitor.log 2>&1"
bind = $mainMod, 9, exec, bash -c "/home/myusername/scripts/hyprland_show_workspace_clean.sh 9 >> /home/myusername/monitor.log 2>&1"
bind = $mainMod, 0, exec, bash -c "/home/myusername/scripts/hyprland_show_workspace_clean.sh 10 >> /home/myusername/monitor.log 2>&1"
[...]
```
## script
```bash
#! /bin/bash
monitors=$(hyprctl monitors | tr '\n' '\a')
activemonitor=$(grep -Po 'Monitor((?!Monitor).)*?focused: yes' <<<"$monitors" | tr '\a' '\n' | awk 'NR==1 {print $2}')
passivemonitor=$(grep -Po 'Monitor((?!Monitor).)*?focused: no' <<<"$monitors" | tr '\a' '\n' | awk 'NR==1 {print $2}')
passivews=$(grep -Po 'Monitor((?!Monitor).)*?focused: no' <<<"$monitors" | tr '\a' '\n' | awk '/active workspace/ {print $3}')


if [[ -z $1 ]]; then
  workspace="$passivews"
  echo "no workspace speficied, use passive monitor workspace: $workspace instead"
else
  echo "workspace: $1 specified"
  workspace="$1"
fi

echo "current layout:"
hyprctl monitors

if [[ $workspace -eq $passivews ]] && [[ $activemonitor != "$passivemonitor" ]]; then
  echo "workspace: $workspace equals passive monitor workspace: $passivews"
  echo "switching to workspace: $workspace"
  hyprctl dispatch workspace "$workspace"
else
  hyprctl dispatch moveworkspacetomonitor "$workspace" "$activemonitor"
  hyprctl dispatch workspace "$workspace"
  echo "moved workspace: $workspace to monitor: $activemonitor"

fi

echo "current layout:"
hyprctl monitors
```

# Expected outcome
When switching to a workspace which is not displayed on any monitor it should open on the currently active monitor

# Noted outcome
Sometimes (I haven't been able to figure out completely why and when it happens, usually i just switch through workspaces 1 through 10 and it is bound to happen at least once) the workspace on the non-focused monitor also switches to a seemingly random one.

### example 1: DP-4 should still be on Workspace 2 but got switched to Workspace 1
```
workspace: 10 specified
current layout:
Monitor DP-3 (ID 1):
 [...]
 active workspace: 9 (9)
 [...]
 focused: yes
 [...]

Monitor DP-4 (ID 2):
 [...]
 active workspace: 2 (2)
 [...]
 focused: no
 [...]


ok
ok
moved workspace: 10 to monitor: DP-3
current layout:
Monitor DP-3 (ID 1):
 [...]
 active workspace: 10 (10)
 [...]
 focused: yes
 [...]

Monitor DP-4 (ID 2):
 [...]
 active workspace: 1 (1)
 [...]
 focused: no
 [...]
```

### example 2: DP-3 should still be on Workspace 3 but got switched to 6
```
workspace: 5 specified
current layout:
Monitor DP-3 (ID 1):
 [...]
 active workspace: 3 (3)
 [...]
 focused: no
 [...]

Monitor DP-4 (ID 2):
 [...]
 active workspace: 4 (4)
 [...]
 focused: yes
 [...]


ok
ok
moved workspace: 5 to monitor: DP-4
current layout:
Monitor DP-3 (ID 1):
 [...]
 active workspace: 6 (6)
 [...]
 focused: no
 [...]

Monitor DP-4 (ID 2):
 [...]
 active workspace: 5 (5)
 [...]
 focused: yes
```

**Comment by vaxerski** [comment]
what hyprland ver? git?

**Comment by macearl** [comment]
I'm currently running the arch package, which is 0.24.1

I previously had the git version from the aur installed, I think I had the issue there as well but I can't be sure at the moment. I will switch to the git version tomorrow and let you know if the issue still exists

**Comment by vaxerski** [comment]
Yeah please do, git had some workspace mechanics overhauls

**Comment by macearl** [comment]
Just switched to the git Package from the AUR
```bash
$  pacman -Q hyprland-git
hyprland-git r2831.72d2f33b-1
```
I'm at a different docking station today, so my outputs are different, but sadly the problem still persists.

Log from my wrapper script: DP-3 should still be on workspace 2 
```
workspace: 1 specified
current layout:
Monitor eDP-1 (ID 0):
 [...]
 active workspace: 3 (3)
 [...]
 focused: yes
 [...]

Monitor DP-3 (ID 1):
 [...]
 active workspace: 2 (2)
 [...]
 focused: no
 [...]


ok
ok
moved workspace: 1 to monitor: eDP-1
current layout:
Monitor eDP-1 (ID 0):
 [...] 
 active workspace: 1 (1)
 [...]
 focused: yes
 [...]

Monitor DP-3 (ID 1):
 [...]
 active workspace: 6 (6)
 [...]
 focused: no
 [...]
```


If there is any kind of additional debug log you need please let me know.

**Comment by fdev31** [comment]
I'm using the git version on arch, to avoid "weird" monitor focus changes and workspace changes I need to move every "inactive" workspaces to the monitor which is being focused.
It works quite well but triggers animations when I change the focused monitor...
Can your problem relate to this one ? When the switch is happening, isn't it because you have the workspace you switch to listed in `hyprctl workspaces` but bound to the other monitor even though not visible ?

**Comment by vaxerski** [comment]
Could it be that
```
hyprctl dispatch moveworkspacetomonitor "$workspace" "$activemonitor"
hyprctl dispatch workspace "$workspace"
```
the order of executing these is somehow flipped? That would explain what is happening. Can you add a sleep 0.1 in between them and check if that fixes the bug?

**Comment by macearl** [comment]
The problem seems to be with
```
hyprctl dispatch moveworkspacetomonitor "$workspace" "$activemonitor"
```

I added a sleep 2 between the two commands to make it extra obvious which command switches the passive monitor and voila the passive monitor switches, 2 second pause, the active monitor switches as well.

@fdev31 seems to be on the right track, the switch seems to only happen if the workspace i switch to was previously on the other Monitor and additionally the active monitor only has the one, currently active workspace assigned.

focused external display, all workspaces on internal display:
Hyprland Log:
```
[LOG] moveWorkspaceToMonitor: Moving 9 to monitor 1
[LOG] moveWorkspaceToMonitor: Plugging gap with existing 3
```
wrapper script log:
```
workspace: 9 specified

current layout:
Monitor DP-3 (ID 1):
 active workspace: 1 (1)
 focused: yes
Monitor eDP-1 (ID 0):
 active workspace: 2 (2)
 focused: no

current workspaces:
workspace ID 1 (1) on monitor DP-3:
workspace ID 3 (3) on monitor eDP-1:
workspace ID 6 (6) on monitor eDP-1:
workspace ID 7 (7) on monitor eDP-1:
workspace ID 10 (10) on monitor eDP-1:
workspace ID 8 (8) on monitor eDP-1:
workspace ID 5 (5) on monitor eDP-1:
workspace ID 9 (9) on monitor eDP-1:
workspace ID 2 (2) on monitor eDP-1:

moved workspace: 9 to monitor: DP-3

current layout:
Monitor DP-3 (ID 1):
 active workspace: 1 (1)
 focused: yes
Monitor eDP-1 (ID 0):
 active workspace: 3 (3)
 focused: no

current workspaces:
workspace ID 1 (1) on monitor DP-3:
workspace ID 3 (3) on monitor eDP-1:
workspace ID 6 (6) on monitor eDP-1:
workspace ID 7 (7) on monitor eDP-1:
workspace ID 10 (10) on monitor eDP-1:
workspace ID 8 (8) on monitor eDP-1:
workspace ID 5 (5) on monitor eDP-1:
workspace ID 9 (9) on monitor DP-3:
workspace ID 2 (2) on monitor eDP-1:
[...]
```

focused internal display, all workspaces on external display:
Hyprland Log:
```
[LOG] moveWorkspaceToMonitor: Moving 1 to monitor 0
[LOG] moveWorkspaceToMonitor: Plugging gap with existing 3
```
wrapper script log:
```
workspace: 1 specified

current layout:
Monitor DP-3 (ID 1):
 active workspace: 2 (2)
 focused: no
Monitor eDP-1 (ID 0):
 active workspace: 6 (6)
 focused: yes

current workspaces:
workspace ID 1 (1) on monitor DP-3:
workspace ID 3 (3) on monitor DP-3:
workspace ID 6 (6) on monitor eDP-1:
workspace ID 7 (7) on monitor DP-3:
workspace ID 10 (10) on monitor DP-3:
workspace ID 8 (8) on monitor DP-3:
workspace ID 5 (5) on monitor DP-3:
workspace ID 9 (9) on monitor DP-3:
workspace ID 2 (2) on monitor DP-3:

moved workspace: 1 to monitor: eDP-1

current layout:
Monitor DP-3 (ID 1):
 active workspace: 3 (3)
 focused: no
Monitor eDP-1 (ID 0):
 active workspace: 6 (6)
 focused: yes

current workspaces:
workspace ID 1 (1) on monitor eDP-1:
workspace ID 3 (3) on monitor DP-3:
workspace ID 6 (6) on monitor eDP-1:
workspace ID 7 (7) on monitor DP-3:
workspace ID 10 (10) on monitor DP-3:
workspace ID 8 (8) on monitor DP-3:
workspace ID 5 (5) on monitor DP-3:
workspace ID 9 (9) on monitor DP-3:
workspace ID 2 (2) on monitor DP-3:
[...]
```

https://github.com/hyprwm/Hyprland/blob/main/src/Compositor.cpp#L1870-L1895
seem to be the corresponding lines of code

**Comment by fdev31** [comment]
You may want to try https://pypi.org/project/pyprland/

Just have ["workspaces_follow_focus"] in the plugins, no configuration needed...

**Comment by dtop129** [comment]
In my build I patched this behaviour by changing the workspace of the old monitor only if the moved workspace was active

```diff
diff --git a/src/Compositor.cpp b/src/Compositor.cpp
index a8764e0e..24399056 100644
--- a/src/Compositor.cpp
+++ b/src/Compositor.cpp
@@ -1892,7 +1892,8 @@ void CCompositor::moveWorkspaceToMonitor(CWorkspace* pWorkspace, CMonitor* pMoni
 
     Debug::log(LOG, "moveWorkspaceToMonitor: Plugging gap with existing %d", nextWorkspaceOnMonitorID);
 
-    POLDMON->changeWorkspace(nextWorkspaceOnMonitorID);
+	if (SWITCHINGISACTIVE)
+		POLDMON->changeWorkspace(nextWorkspaceOnMonitorID);
 
     // move the workspace
 

```

**Comment by vaxerski** [comment]
right, try this
[patch.txt](https://github.com/hyprwm/Hyprland/files/11354605/patch.txt)

**Comment by macearl** [comment]
That seems to have fixed it for me

**Comment by vaxerski** [comment]
fixed in fbb938fcf228b84dc82b5ad4f85387448d1a2d9f then

**Comment by fdev31** [comment]
Also fixed https://github.com/hyprland-community/pyprland/issues/1 :)
Good job :+1:

**Comment by dtop129** [comment]
In the moveworkspacetomonitor function it would also be nice if, when moving an active workspace(not focused) to the current monitor, the workspace on the focused monitor didn't change.
This would also fix issue #1980

```diff
diff --git a/src/Compositor.cpp b/src/Compositor.cpp
index 4222735a..414020b5 100644
--- a/src/Compositor.cpp
+++ b/src/Compositor.cpp
@@ -1925,7 +1925,7 @@ void CCompositor::moveWorkspaceToMonitor(CWorkspace* pWorkspace, CMonitor* pMoni
         }
     }
 
-    if (SWITCHINGISACTIVE) { // if it was active, preserve its' status. If it wasn't, don't.
+    if (SWITCHINGISACTIVE && POLDMON == g_pCompositor->m_pLastMonitor) { // if it was focused, preserve its' status. If it wasn't, don't.
         Debug::log(LOG, "moveWorkspaceToMonitor: SWITCHINGISACTIVE, active %d -> %d", pMonitor->activeWorkspace, pWorkspace->m_iID);
 
         if (const auto PWORKSPACE = getWorkspaceByID(pMonitor->activeWorkspace); PWORKSPACE)

```

**Comment by vaxerski** [comment]
@dtop129 applied :)
